ssl/ssl_lib.c

```diff
@@ -2205,20 +2212,41 @@ long SSL_CTX_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_SESS_CACHE_SIZE:
         if (larg < 0)
             return 0;
-        l = (long)ctx->session_cache_size;
-        ctx->session_cache_size = (size_t)larg;
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = (long)ctx->session_cache->size;
+        ctx->session_cache->size = (size_t)larg;
+        CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+        CRYPTO_THREAD_unlock(ctx->lock);
         return l;
     case SSL_CTRL_GET_SESS_CACHE_SIZE:
-        return (long)(ctx->session_cache_size);
+        CRYPTO_THREAD_read_lock(ctx->lock);
+        l = (long)ctx->session_cache->size;
+        CRYPTO_THREAD_unlock(ctx->lock);
+        return l;
+
     case SSL_CTRL_SET_SESS_CACHE_MODE:
-        l = ctx->session_cache_mode;
-        ctx->session_cache_mode = larg;
-        return (l);
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = ctx->session_cache->mode;
+        ctx->session_cache->mode = larg;
```

Comment:
If two SSL_CTX's are sharing a cache, they need to use the same mode; so it makes sense that the mode (and timeout, etc) are properties of the session_cache.

> Maybe the mode should not be changeable after a cache is assigned to a session.

Not clear on the comment here, a cache is assigned to an SSL_CTX at initialization, in such a case, this would prevent the mode from being changed. Unless you mean that a session is assigned to a cache, which makes sense, but is no different than now.